app/master/build.py

```diff
@@ -445,18 +450,25 @@ def _status(self):
         else:
             return BuildStatus.BUILDING
 
-    def _failed_atoms(self):
+    def failed_atoms(self):
         """
-        The commands which failed
-        :rtype: list [str] | None
+        The atoms that failed. Returns None if the build hasn't completed yet. Returns empty set if
+        build has completed and no atoms have failed.
+        :rtype: list[Atom] | None
         """
-        if self._is_canceled:
-            return []
+        if self._failed_atoms is None and self.is_finished:
+            # Why was this conditional here in the first place? Don't we only care about whether or not
+            # the build is finished? I think we can remove this.
```

Comment:
Are you wondering about the `_is_canceled` check? It would be awesome to remove the check but obvs we should do a manual test first. I wonder if build_artifact was raising an exception since a canceled build doesn't have an artifact dir?